### PR TITLE
machine: add vmwarefusion support

### PR DESCRIPTION
**Pull request by philips**
```
$ ./out/minikube start --help
Starts a local kubernetes cluster using Virtualbox. This command
assumes you already have Virtualbox installed.

Usage:
  minikube start [flags]

Flags:
      --iso-url string     Location of the minikube iso (default "https://storage.googleapis.com/minikube/minikube-0.3.iso")
      --vm-driver string   VM driver is either vmwarefusion or virtualbox (default "virtualbox")
```

```
./out/minikube start --vm-driver=vmwarefusion
```

### CODE REVIEW

**Comment by philips** [comment]
Anyone have some debugging tips?

**Comment by dlorenc** [comment]
Hey,

I don't see anything obviously wrong with your change, it looks like the right shape overall. I'll grab a copy of vmware fusion and see if I can figure anything out.

One guess (this might be a documentation bug) is that the vmware fusion docs mention a different boot2docker iso: https://docs.docker.com/machine/drivers/vm-fusion/, specifically this repo: https://github.com/frapposelli/boot2docker/tree/vmware-64bit

which has some changes that don't seem to be upstreamed.

**Comment by philips** [comment]
@dlorenc LMK if you need a license key for VMWare Fusion. Happy to pay for your copy to help me out :)

**Comment by dlorenc** [comment]
I tried starting up with your patch and saw this in ~/.minikube/machines/minikubeVM/vmware-0.log file:

```shell
2016-06-08T09:35:45.463-08:00| vmx| I125: MsgIsAnswered: Using builtin default 'OK' as the answer for 'msg.vmx.noVCPUsConfigured'
2016-06-08T09:35:45.463-08:00| vmx| I125: Module 'NumVCPUs' power on failed.
```

The minikubeVM.vmx file generated by libmachine has this line in it: ```numvcpus = "0"```

The template file [here](https://github.com/docker/machine/blob/master/drivers/vmwarefusion/vmx_darwin.go#L69) tries to reference ".CPU", but that driver appears to have both ".CPUS" and ".CPU" fields, and I think the constructor you were using only sets one: https://github.com/docker/machine/blob/master/drivers/vmwarefusion/fusion_darwin.go#L119

I haven't gotten it working yet, but I think it's just a mismatch between what parameters get initialized by NewDriver and SetConfigFromFlags. For the virtualbox driver they seem to use the same defaults everywhere, but in fusion they don't.

**Comment by dlorenc** [comment]
This commit here seems to work, and I got a cluster running!
https://github.com/dlorenc/minikube/commit/16e323eddedb0d248fc0a265aca4b1b5f43e1f85

I think the fix should probably go upstream into the NewDriver function in machine, though.

**Comment by philips** [comment]
@dlorenc THANKS! Any preference on what the driver flag is called?

**Comment by luxas** [comment]
--vm-driver maybe?

BTW, we've talked about

--experimental-provider 

which will turn into --localkube-provider or something when it's stable

It's for enabling new providers like the host-native mode that's in our roadmap

**Comment by dlorenc** [comment]
I guess there are a few levels of configuration we're discussing here:

1. Whether to use libmachine to provision a VM at all (vs. running natively on Linux or inside a BYO Docker setup)
2. Which libmachine driver we want to use (virtualbox, hyper-v, vmware fusion, etc.)

Then we also have the question of how to pass driver-specific configuration to the libmachine drivers (memory, cpu, etc.)

I'd like to avoid cloning the entire machine CLI parameter set, but we will need some way to configure these different levels. I'll think about it a little more.

**Comment by philips** [comment]
OK, I am going to go with `--vm-driver`.

**Comment by philips** [comment]
@dlorenc Alright! Got it all working and plumed. PTAL!

**Comment by philips** [comment]
@luxas I ended up going with vm-driver, it all works now

File: pkg/minikube/cluster/cluster_test.go [hunk 1]
```diff
@@ -57,8 +57,8 @@ func TestCreateHost(t *testing.T) {
 		t.Fatalf("Machine is not running. State is: %s", s)
 	}
 
-	if h.DriverName != "virtualbox" {
-		t.Fatalf("Wrong driver name: %v. Should be virtualbox.", h.DriverName)
+	if h.DriverName != "virtualbox" || h.DriverName != "vmwarefusion" {
```
Comment (luxas) [inline pkg/minikube/cluster/cluster_test.go]: I'd like if you made an array of supported drivers that could be used when validating the flag and here in the test

File: README.md [hunk 2]
```diff
@@ -23,7 +23,7 @@ Here's a brief demo of minikube usage. We're using the code from this [Kubernete
 Note that the IP below is dynamic and can change. It can be retrieved with `minikube ip`.
 
 ```shell
-$ minikube start
+$ minikube start --vm-driver=virtualbox
```
Comment (luxas) [inline README.md]: Maybe just leave this out for a note that vbox is the default?

File: pkg/minikube/cluster/cluster_test.go (hunk 1 above)
Comment (philips) [inline pkg/minikube/cluster/cluster_test.go]: sure, np

File: README.md (hunk 2 above)
Comment (philips) [inline README.md]: @luxas OK, I can do that.

File: pkg/minikube/cluster/cluster_test.go [hunk 3]
```diff
@@ -57,8 +57,15 @@ func TestCreateHost(t *testing.T) {
...
 
-	if h.DriverName != "virtualbox" {
-		t.Fatalf("Wrong driver name: %v. Should be virtualbox.", h.DriverName)
+	for _, driver := range constants.SupportVMDrivers {
+		found := false
+		if h.DriverName == driver {
+			found = true
+		}
+
+		if !found {
```
Comment (dlorenc) [inline pkg/minikube/cluster/cluster_test.go]: Should this be outside the loop?

Comment (philips) [inline pkg/minikube/cluster/cluster_test.go]: @dlorenc wow, embarrassing, fixed :(

File: pkg/minikube/cluster/cluster.go [hunk 4]
```diff
@@ -225,15 +227,32 @@ func SetupCerts(d drivers.Driver) error {
...
-	driver.Boot2DockerURL = config.MinikubeISO
+	var driver drivers.Driver
+
+	switch config.VMDriver {
+	case "virtualbox":
+		d := virtualbox.NewDriver(constants.MachineName, constants.Minipath)
+		d.Boot2DockerURL = config.MinikubeISO
+		driver = d
+	case "vmwarefusion":
+		d := vmwarefusion.NewDriver(constants.MachineName, constants.Minipath).(*vmwarefusion.Driver)
```
Comment (dlorenc) [inline pkg/minikube/cluster/cluster.go]: Looks like the build is still failing:  `pkg/minikube/cluster/cluster.go:274: undefined: vmwarefusion.Driver`

I think this should be `.(*drivers.Driver)`

**Comment by philips** [comment]
@dlorenc Finally got this thing building on travis. The tl;dr is that the vmwarefusion driver only compiles on Darwin.

**Comment by aaron-prindle** [comment]
LGTM

**Pull request merged**